### PR TITLE
fix: schema not cached in spark/flink loader

### PR DESCRIPTION
**Pull request by haohao0103**
SchemaCache init and after query schema then put the latest schema to SchemaCache

I have read the CLA Document and I hereby sign the CLA

### CODE REVIEW

**Comment by haohao0103** [comment]
fix #333
Fix bug: SchemaCache init and after query schema then put the latest schema to SchemaCache

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/builder/SchemaCache.java [hunk 1]
```diff
@@ -46,6 +46,7 @@ public SchemaCache(HugeClient client) {
         this.propertyKeys = new HashMap<>();
         this.vertexLabels = new HashMap<>();
         this.edgeLabels = new HashMap<>();
+        updateAll();
```
Comment (imbajin) [inline hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/builder/SchemaCache.java:49]: why we need clear the cache when we firstly create it?

**Comment by imbajin** [comment]
2 little questions:
1. why it only influence spark-loader now?
2. shall we flush the cache periodically?

File: hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/builder/SchemaCache.java (hunk 1 above)
Comment (haohao0103) [reply hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/builder/SchemaCache.java:49]: The HugeGraphSparkLoader does not get the real metadata information when SchemaCache is first created

**Comment by haohao0103** [comment]
> 2 little questions:
> 
> 1. why it only influences `spark-loader` now?
> 2. shall we flush the cache periodically?

1、Hugegraphloader中load()方法会执行this.context.updateSchemaCache()更新元数据
2、可以改成第一次的初始化，或者改成第一次查询后更新，这两种我觉得保留一种即可

**Comment by simon824** [comment]
hi @haohao0103 ,thanks for your contribution.
I think maybe it's better to updateSchemaCache after `new LoadContext` just like `HugeGraphLoader`, what do you think? @imbajin @haohao0103 
https://github.com/apache/incubator-hugegraph-toolchain/blob/master/hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/spark/HugeGraphSparkLoader.java#L120

``` java
    private LoadContext initPartition(
            LoadOptions loadOptions, InputStruct struct) {
        LoadContext context = new LoadContext(loadOptions);
        for (VertexMapping vertexMapping : struct.vertices()) {
            this.builders.put(
                    new VertexBuilder(context, struct, vertexMapping),
                    new ArrayList<>());
        }
        for (EdgeMapping edgeMapping : struct.edges()) {
            this.builders.put(new EdgeBuilder(context, struct, edgeMapping),
                              new ArrayList<>());
        }
        context.updateSchemaCache();
        return context;
    }
```

**Comment by haohao0103** [comment]
> hi @haohao0103 ,thanks for your contribution. I think maybe it's better to updateSchemaCache after `new LoadContext` just like `HugeGraphLoader`, what do you think? @imbajin @haohao0103 https://github.com/apache/incubator-hugegraph-toolchain/blob/master/hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/spark/HugeGraphSparkLoader.java#L120
> 
> ```java
>     private LoadContext initPartition(
>             LoadOptions loadOptions, InputStruct struct) {
>         LoadContext context = new LoadContext(loadOptions);
>         for (VertexMapping vertexMapping : struct.vertices()) {
>             this.builders.put(
>                     new VertexBuilder(context, struct, vertexMapping),
>                     new ArrayList<>());
>         }
>         for (EdgeMapping edgeMapping : struct.edges()) {
>             this.builders.put(new EdgeBuilder(context, struct, edgeMapping),
>                               new ArrayList<>());
>         }
>         context.updateSchemaCache();
>         return context;
>     }
> ```

我觉得放在这里是可以的，我们第一次就在这里解决的；但是initPartition方法是HugeGraphSparkLoader类私有的；感觉LoadContext对象创建后需要调用initPartition初始化方法才能正确使用，感觉有点不合适哈。个人的浅见。。。。

**Comment by simon824** [comment]
> > hi @haohao0103 ,thanks for your contribution. I think maybe it's better to updateSchemaCache after `new LoadContext` just like `HugeGraphLoader`, what do you think? @imbajin @haohao0103 https://github.com/apache/incubator-hugegraph-toolchain/blob/master/hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/spark/HugeGraphSparkLoader.java#L120
> > ```java
> >     private LoadContext initPartition(
> >             LoadOptions loadOptions, InputStruct struct) {
> >         LoadContext context = new LoadContext(loadOptions);
> >         for (VertexMapping vertexMapping : struct.vertices()) {
> >             this.builders.put(
> >                     new VertexBuilder(context, struct, vertexMapping),
> >                     new ArrayList<>());
> >         }
> >         for (EdgeMapping edgeMapping : struct.edges()) {
> >             this.builders.put(new EdgeBuilder(context, struct, edgeMapping),
> >                               new ArrayList<>());
> >         }
> >         context.updateSchemaCache();
> >         return context;
> >     }
> > ```
> 
> 我觉得放在这里是可以的，我们第一次就在这里解决的；但是initPartition方法是HugeGraphSparkLoader类私有的；感觉需要调用initPartition方法才能正确正确创建及初始化LoadContext对象，感觉有点不合适哈。个人的浅见。。。。

没有明白你的意思，你是指哪里不合适？LoadContext 应该是只有在初始化分区的时候才需要创建？

**Comment by haohao0103** [comment]
> > > hi @haohao0103 ,thanks for your contribution. I think maybe it's better to updateSchemaCache after `new LoadContext` just like `HugeGraphLoader`, what do you think? @imbajin @haohao0103 https://github.com/apache/incubator-hugegraph-toolchain/blob/master/hugegraph-loader/src/main/java/com/baidu/hugegraph/loader/spark/HugeGraphSparkLoader.java#L120
> > > ```java
> > >     private LoadContext initPartition(
> > >             LoadOptions loadOptions, InputStruct struct) {
> > >         LoadContext context = new LoadContext(loadOptions);
> > >         for (VertexMapping vertexMapping : struct.vertices()) {
> > >             this.builders.put(
> > >                     new VertexBuilder(context, struct, vertexMapping),
> > >                     new ArrayList<>());
> > >         }
> > >         for (EdgeMapping edgeMapping : struct.edges()) {
> > >             this.builders.put(new EdgeBuilder(context, struct, edgeMapping),
> > >                               new ArrayList<>());
> > >         }
> > >         context.updateSchemaCache();
> > >         return context;
> > >     }
> > > ```
> > 
> > 
> > 我觉得放在这里是可以的，我们第一次就在这里解决的；但是initPartition方法是HugeGraphSparkLoader类私有的；感觉需要调用initPartition方法才能正确正确创建及初始化LoadContext对象，感觉有点不合适哈。个人的浅见。。。。
> 
> 没有明白你的意思，你是指哪里不合适？LoadContext 应该是只有在初始化分区的时候才需要创建？

我的意思是这样会不会限定了LoadContext对象只有在HugeGraphSparkLoader类才能被正确初始化，限制了扩展性；我们在开发Bulkload代码的时候，需要扩展新的类然后使用LoadContext对象，就没办法调用initPartition方法了；在另外的类中构建LoadContext对象时需要再显示执行下updateSchemaCache()方法才能保证loadcontext对象可以正确使用。这个是我的理解哈，代码设计等我也不是很精通，仅供大佬参考哈

**Comment by simon824** [comment]
@haohao0103 明白你的意思了，但是 `HugeGraphLoader` 还有一个创建 Schema 的逻辑`createSchema`，SparkLoader 还没有加这个逻辑，但是也是需要的，创建 schema 后再加载 schema 缓存看起来更合理一些。另外如果有必要的话完全可以将 initPartition 方法改为 public，或者定义一个其他更通用的方法。

**Comment by haohao0103** [comment]
> @haohao0103 明白你的意思了，但是 `HugeGraphLoader` 还有一个创建 Schema 的逻辑`createSchema`，SparkLoader 还没有加这个逻辑，但是也是需要的，创建 schema 后再加载 schema 缓存看起来更合理一些。另外如果有必要的话完全可以将 initPartition 方法改为 public，或者定义一个其他更通用的方法。

明白了，谢谢哈

**Comment by haohao0103** [comment]
> > @haohao0103 明白你的意思了，但是 `HugeGraphLoader` 还有一个创建 Schema 的逻辑`createSchema`，SparkLoader 还没有加这个逻辑，但是也是需要的，创建 schema 后再加载 schema 缓存看起来更合理一些。另外如果有必要的话完全可以将 initPartition 方法改为 public，或者定义一个其他更通用的方法。
> 
> 明白了，谢谢哈

我需要重新提交一个pr?

**Comment by simon824** [comment]
@haohao0103 可以直接在这个PR上提交代码覆盖

**Comment by haohao0103** [comment]
> @haohao0103 可以直接在这个PR上提交代码覆盖

已提交代码

**Review by simon824 — APPROVED** [review]

**Review by imbajin — DISMISSED** [review]
记得也顺便检查一下 `flink-loader` 那边是否也需要加一下

如果需要 public 的话可以改

**Comment by haohao0103** [comment]
> > 记得也顺便检查一下 `flink-loader` 那边是否也需要加一下
> > 如果需要 public 的话可以改
> 
> 需要修改，需要在initBuilders()方法中调用context.updateSchemaCache();

 private Map<ElementBuilder, List<String>> initBuilders() {
        LoadContext loadContext = new LoadContext(this.loadOptions);
        Map<ElementBuilder, List<String>> builders = new HashMap<>();
        for (VertexMapping vertexMapping : this.struct.vertices()) {
            builders.put(new VertexBuilder(loadContext, this.struct, vertexMapping),
                         new ArrayList<>());
        }
        for (EdgeMapping edgeMapping : this.struct.edges()) {
            builders.put(new EdgeBuilder(loadContext, this.struct, edgeMapping),
                         new ArrayList<>());
        }
// TODO       
//  loadContext.updateSchemaCache();
        return builders;
    }

**Comment by haohao0103** [comment]
> 记得也顺便检查一下 `flink-loader` 那边是否也需要加一下
> 
> 如果需要 public 的话可以改

还可以继续提代码吗？

**Comment by haohao0103** [comment]
> > 记得也顺便检查一下 `flink-loader` 那边是否也需要加一下
> > 如果需要 public 的话可以改
> 
> 还可以继续提代码吗？

@imbajin @simon824 flink-loader的代码也修复了，麻烦review。谢谢

**Review by simon824 — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Review by imbajin — APPROVED** [review]

**Pull request merged**